lib/grant-types/abstract-grant-type.ts

```diff
@@ -0,0 +1,112 @@
+import { MILLISECONDS_PER_SECOND } from '../constants';
+import { InvalidArgumentError, InvalidScopeError } from '../errors';
+import { Client, Model, User } from '../interfaces';
+import { Request } from '../request';
+import * as tokenUtil from '../utils/token-util';
+import * as is from '../validator/is';
+
+export class AbstractGrantType {
+  accessTokenLifetime: number;
+  model: Model;
+  refreshTokenLifetime: number;
+  alwaysIssueNewRefreshToken: boolean;
+
+  constructor(options: any = {}) {
+    if (!options.accessTokenLifetime) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `accessTokenLifetime`',
+      );
+    }
+
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    this.accessTokenLifetime = options.accessTokenLifetime;
+    this.model = options.model;
+    this.refreshTokenLifetime = options.refreshTokenLifetime;
+    this.alwaysIssueNewRefreshToken = options.alwaysIssueNewRefreshToken;
+  }
+
+  /**
+   * Generate access token.
+   */
+
+  async generateAccessToken(client?: Client, user?: User, scope?: string) {
+    if (this.model.generateAccessToken) {
+      const token = await this.model.generateAccessToken(client, user, scope);
+
+      return token ? token : tokenUtil.GenerateRandomToken();
+    }
+
+    return tokenUtil.GenerateRandomToken();
+  }
+
+  /**
+   * Generate refresh token.
+   */
+
+  async generateRefreshToken(client?: Client, user?: User, scope?: string) {
+    if (this.model.generateRefreshToken) {
+      const token = await this.model.generateRefreshToken(client, user, scope);
+
+      return token ? token : tokenUtil.GenerateRandomToken();
+    }
+
+    return tokenUtil.GenerateRandomToken();
+  }
+
+  /**
+   * Get access token expiration date.
+   */
+
+  getAccessTokenExpiresAt() {
+    return new Date(
+      Date.now() + this.accessTokenLifetime * MILLISECONDS_PER_SECOND,
+    );
+  }
+
+  /**
+   * Get refresh token expiration date.
+   */
+
+  getRefreshTokenExpiresAt() {
+    return new Date(
+      Date.now() + this.refreshTokenLifetime * MILLISECONDS_PER_SECOND,
+    );
+  }
+
+  /**
+   * Get scope from the request body.
+   */
+
+  getScope(request: Request) {
+    if (!is.nqschar(request.body.scope)) {
+      throw new InvalidArgumentError('Invalid parameter: `scope`');
+    }
+
+    return request.body.scope;
+  }
+
+  /**
+   * Validate requested scope.
+   */
+  async validateScope(user: User, client: Client, scope: string) {
+    if (this.model.validateScope) {
+      const validatedScope = await this.model.validateScope(
+        user,
+        client,
+        scope,
+      );
+      if (!validatedScope) {
+        throw new InvalidScopeError(
+          'Invalid scope: Requested scope is invalid',
+        );
+      }
+
+      return validatedScope;
+    }
+
+    return scope;
```

Comment:
If `this.model.validateScope` is false, we just return the passed in scope; unless I'm reading it wrong.

I'm assuming that's expected behaviour? (I suppose one could call a validate scope on a model that simply does not validate scopes but it does seem odd to my mind, at least).

Reply:
this method is in an abstract class which provides methods for other classes, that inherit from this class (basic concept in OOP). If the method does not need to validateScope then there is no scope validation happening and you get your original scope.